### DISCUSSION TITLE
Phoenix Playground: Single File Admin Panel?

### DISCUSSION THREAD

**Discussion by pehbehbeh** [naymspace/backpex#412] [Ideas]
Dashbit recently [announced](https://dashbit.co/blog/announcing-phoenix-playground) [Phoenix Playground](https://github.com/phoenix-playground/phoenix_playground).

In this discussion I would like to evaluate if there is an easy way to provide a single file admin panel with Backpex.

**Comment by pehbehbeh** [comment]
Hey @wojtekmach, I talked to @krns and your suggestion to implement a single file admin panel with Phoenix Playground and Backpex. Great idea!

There are a few questions that come to mind and I thought it would be best to discuss them here.

We depend on Tailwind, daisyUI and Alpine.js. Maybe we could use the CDN versions of the tools for this use case. It won't be used in production anyway (or will it)? But I'm not sure if that would work well with LiveView. Something like that:

```
<link href="https://cdn.jsdelivr.net/npm/daisyui@4.12.8/dist/full.min.css" rel="stylesheet" type="text/css" />
<script src="https://cdn.tailwindcss.com"></script>
<script defer src="https://cdn.jsdelivr.net/npm/alpinejs@3.x.x/dist/cdn.min.js"></script>
```

Is it possible to provide a base layout?

How would you handle migrations?

**Comment by pehbehbeh** [comment]
Another challenge will be the routing. Currently we need to add some macro stuff to the router and Backpex generates a lot of routes based on that. I saw that you can provide a `:plug` to [`PhoenixPlayground.start/1`](https://hexdocs.pm/phoenix_playground/0.1.3/PhoenixPlayground.html#start/1). So maybe we can add a minimal router module.

**Comment by wojtekmach** [comment]
Regarding layout and CSS/JS, I have a few open issues https://github.com/phoenix-playground/phoenix_playground/issues and I'm still considering options but I am leaning to something simple along the lines of:

```elixir
PhoenixPlayground.start(
  live: DemoLive,
  extra_head_html: """
  <script ...>
  """
)
```

If it's more practical to replace the layout wholesale, there could be a `:layout` option too. At the moment probably the best option is to use workaround described here: https://github.com/phoenix-playground/phoenix_playground/issues/2#issuecomment-2112975557.

Regarding router, I could add support for defining multiple routes but if it is more practical to bring your own router, `:plug` option is precisely for that.

Regarding migrations, you can check out this script:

https://github.com/wojtekmach/mix_install_examples/blob/89e9612028d5f157a672194bbe3550a2471195cd/ecto_sql.exs#L62

In that script we always re-migrate the DB which would most likely not be a good idea here.

You can start Repo in the same supervision tree as the Playground endpoint by using `:child_specs` option.

So I think all the foundational pieces are here already, perhaps?
